src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Cohost/RazorCohostCapabilitiesProvider.cs

```diff
@@ -0,0 +1,33 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using Microsoft.AspNetCore.Razor.LanguageServer;
+using Microsoft.CodeAnalysis.ExternalAccess.Razor.Cohost;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Newtonsoft.Json;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor.Cohost;
+
+[Export(typeof(IRazorCohostCapabilitiesProvider)), System.Composition.Shared]
+[method: ImportingConstructor]
+internal class RazorCohostCapabilitiesProvider([ImportMany(typeof(ICapabilitiesProvider))] IEnumerable<Lazy<ICapabilitiesProvider>> cohostCapabilitiesProviders) : IRazorCohostCapabilitiesProvider
+{
+    private readonly IEnumerable<Lazy<ICapabilitiesProvider>> _cohostCapabilitiesProviders = cohostCapabilitiesProviders;
+
+    public string GetCapabilities(string clientCapabilities)
+    {
+        var clientCapabilitiesObject = JsonConvert.DeserializeObject<VSInternalClientCapabilities>(clientCapabilities) ?? new();
+
+        var serverCapabilities = new VSInternalServerCapabilities();
+
+        foreach (var provider in _cohostCapabilitiesProviders)
```

Comment:
Our capabilities are not any difference right? This is just needed so we can run both cohosting and our current lsp server together?

Reply:
No, the capabilities are different: We don't offer the same capability in cohosting and non. They sum together to the full set we support.

This is _the_ place for capabilities to be set in cohosting, and some version of this code will stay. The way it works is that endpoints apply their own capabilities, so as we expose an endpoint to cohosting, via MEF, we remove it from the normal server, but not registering it in DI.